Reachable unwrap panic in `ProcessFilter::from_which_and_id()`

### Describe the bug
There is a reachable unwrap panic in `ProcessFilter::from_which_and_id()` at kernel/aster-nix/src/process/process_filter.rs:24 when make a `waitid` syscall.

https://github.com/asterinas/asterinas/blob/ce2af1eb057077753a7a757edc1833e677a83918/kernel/aster-nix/src/process/process_filter.rs#L17-L26


### To Reproduce
1. Compile a program which calls `waitid`:
```C
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/wait.h>

int main() {
  waitid(0x31142aad0bc93209, 15, 0x1, -10);

  return 0;
}

```
2. Run the compiled program in Asterinas.

### Expected behavior
Asterinas reports panic and is terminated.

### Environment
- Official docker asterinas/asterinas:0.6.2
- 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
- Asterinas version: main ce2af1eb

### Logs
```
~ # /root/waitpid.c
panicked at /root/asterinas/kernel/aster-nix/src/process/process_filter.rs:24:18:
Unknown id type
Printing stack trace:
   1: fn 0xffffffff8876b840 - pc 0xffffffff8876b858 / registers:

     rax               0x12; rdx 0xffffffff889fd930; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80000122a370;

   2: fn 0xffffffff8876b620 - pc 0xffffffff8876b798 / registers:

     rax               0x12; rdx 0xffffffff889fd930; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80000122a380;

   3: fn 0xffffffff88049000 - pc 0xffffffff8804900a / registers:

     rax               0x12; rdx 0xffffffff889fd930; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80000122a500;

   4: fn 0xffffffff8897ee20 - pc 0xffffffff8897eea2 / registers:

     rax               0x12; rdx 0xffffffff889fd930; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80000122a510;

   5: fn 0xffffffff88163050 - pc 0xffffffff881630b0 / registers:

     rax               0x12; rdx 0xffffffff889fd930; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80000122a5a0;

   6: fn 0xffffffff8822cc20 - pc 0xffffffff8822cc8d / registers:

     rax               0x12; rdx 0xffffffff889fd930; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80000122a610;

   7: fn 0xffffffff88292390 - pc 0xffffffff882bebda / registers:

     rax               0x12; rdx 0xffffffff889fd930; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80000122a730;

   8: fn 0xffffffff88291f70 - pc 0xffffffff88291ffe / registers:

     rax               0x12; rdx 0xffffffff889fd930; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff8000012403d0;

   9: fn 0xffffffff8836de80 - pc 0xffffffff8836e9ff / registers:

     rax               0x12; rdx 0xffffffff889fd930; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff800001240570;

  10: fn 0xffffffff885a49f0 - pc 0xffffffff885a49fe / registers:

     rax               0x12; rdx 0xffffffff889fd930; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff800001240f90;

  11: fn 0xffffffff88781b90 - pc 0xffffffff88781ba6 / registers:

     rax               0x12; rdx 0xffffffff889fd930; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff800001240fb0;

  12: fn 0xffffffff887b9740 - pc 0xffffffff887b97a9 / registers:

     rax               0x12; rdx 0xffffffff889fd930; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff800001240fd0;

  13: fn                0x0 - pc                0x0 / registers:

     rax               0x12; rdx 0xffffffff889fd930; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff800001241000;

[OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) /root/asterinas/ostd/src/panicking.rs:107
(  2) /root/asterinas/ostd/src/panicking.rs:59
(  3) 2aghao2n2kcoquuybdyjuveav:?
(  4) ??:?
(  5) /root/asterinas/kernel/aster-nix/src/process/process_filter.rs:21
(  6) /root/asterinas/kernel/aster-nix/src/syscall/waitid.rs:18
(  7) /root/asterinas/kernel/aster-nix/src/syscall/mod.rs:195
(  8) /root/asterinas/kernel/aster-nix/src/syscall/mod.rs:325
(  9) /root/asterinas/kernel/aster-nix/src/thread/task.rs:69
( 10) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:79
( 11) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2077
( 12) /root/asterinas/ostd/src/task/task/mod.rs:310
make: *** [Makefile:153: run] Error 1
```

--- a/kernel/aster-nix/src/process/process_filter.rs
+++ b/kernel/aster-nix/src/process/process_filter.rs
@@ -14,14 +14,15 @@ pub enum ProcessFilter {
 
 impl ProcessFilter {
     // used for waitid
-    pub fn from_which_and_id(which: u64, id: u64) -> Self {
+    pub fn from_which_and_id(which: u64, id: u64) -> Result<Self> {
         // Does not support PID_FD now(which = 3)
         // https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/wait.h#L20
         match which {
-            0 => ProcessFilter::Any,
-            1 => ProcessFilter::WithPid(id as Pid),
-            2 => ProcessFilter::WithPgid(id as Pgid),
-            _ => panic!("Unknown id type"),
+            0 => Ok(ProcessFilter::Any),
+            1 => Ok(ProcessFilter::WithPid(id as Pid)),
+            2 => Ok(ProcessFilter::WithPgid(id as Pgid)),
+            3 => todo!(),
+            _ => return_errno_with_message!(Errno::EINVAL, "invalid which"),
         }
     }
 

--- a/kernel/aster-nix/src/syscall/waitid.rs
+++ b/kernel/aster-nix/src/syscall/waitid.rs
@@ -16,7 +16,8 @@ pub fn sys_waitid(
 ) -> Result<SyscallReturn> {
     // FIXME: what does infoq and rusage use for?
     let process_filter = ProcessFilter::from_which_and_id(which, upid)?;
-    let wait_options = WaitOptions::from_bits(options as u32).expect("Unknown wait options");
+    let wait_options = WaitOptions::from_bits(options as u32)
+        .ok_or(Error::with_message(Errno::EINVAL, "invalid options"))?;
     let waited_process = wait_child_exit(process_filter, wait_options)?;
     let pid = waited_process.map_or(0, |process| process.pid());
     Ok(SyscallReturn::Return(pid as _))

--- a/kernel/aster-nix/src/syscall/waitid.rs
+++ b/kernel/aster-nix/src/syscall/waitid.rs
@@ -15,7 +15,7 @@ pub fn sys_waitid(
     _ctx: &Context,
 ) -> Result<SyscallReturn> {
     // FIXME: what does infoq and rusage use for?
-    let process_filter = ProcessFilter::from_which_and_id(which, upid);
+    let process_filter = ProcessFilter::from_which_and_id(which, upid)?;
     let wait_options = WaitOptions::from_bits(options as u32).expect("Unknown wait options");
     let waited_process = wait_child_exit(process_filter, wait_options)?;
     let pid = waited_process.map_or(0, |process| process.pid());

